### ISSUE TITLE
idea for GixSQL: provide a GnuCOBOL specific version

### ISSUE THREAD

**Issue by GitMensch** [mridoni/gix#38]
This idea is definitely for "later" - but I want to suggest it now to allow discussions on this idea (which may also lead to GnuCOBOL's COBOL<->C API to be extended).

The GnuCOBOL X specific version would have the same "base code" but would not be usable any more with old versions of OpenCOBOL/GnuCOBOL or any other COBOL (which I think GixSQL currently is working on), which I consider a big downside.

The big benefits would be:
* faster preparsing: the preparser does not need to check much of the COBOL variable
* smaller generated COBOL sources
* therefore also bit faster compiles and smaller binaries
* faster runtime and guaranteed correct values: libcob will read and write those

The clean way to have both would be to use conditional compilation on the sources so GixIDE and other users can have the one that matches their environment best. The actual functions generated should likely be duplicated and renamed so it is directly visible which function is used.

It would change the logic (pseudocode) from
COBOL<->GIXSQL-Entries<->(->store_parameters->)read_write_values_from_COBOL<->db
to
COBOL<->GIXSQL-Entries<->(->get_cob_field->)let_libcob_read_write_values_from_cob_field<->db

The interface in COBOL would change from

```cobol
GIXSQL     CALL "GIXSQLSetSQLParams" USING
GIXSQL          BY VALUE 9
GIXSQL          BY VALUE 18
GIXSQL          BY VALUE -3
GIXSQL          BY REFERENCE AMNT
GIXSQL     END-CALL
```

to plain

```cobol
GIXSQL     CALL "GIXSQLSetSQLParamsGC" USING AMNT
GIXSQL     END-CALL
```

`GIXSQLSetSQLParamsGC` would _ignore_ its parameter `(void *dummy)` and call `cob_get_param_field(1)` to get the `*cob_field` and store this reference (`#include <libcob.h>` for that).
Then it may call any of the following functions to do whatever it needs to do at any later time:

```c
COB_EXPIMP int		cob_get_field_size (const cob_field *);
COB_EXPIMP int		cob_get_field_type (const cob_field *);
COB_EXPIMP int		cob_get_field_digits	(const cob_field *); // 0 means "not numeric"
COB_EXPIMP int		cob_get_field_scale	(const cob_field *);
COB_EXPIMP int		cob_get_field_sign	(const cob_field *);
COB_EXPIMP int		cob_get_field_constant (const cob_field *); // "1 means "may only be read"
```

As most DB functions seem to operate with a "C string" the following functions _may_ be the only ones (or most ones) used:

```c
/* get the field's pretty-display value */
COB_EXPIMP const char	*cob_get_field_str (const cob_field *, char *buff, size_t size);
/* get the field's pretty-display value with an internal buffer for one-time access */
COB_EXPIMP const char	*cob_get_field_str_buffered (const cob_field *);
/* set the field's data using the appropriate internal type, returns EINVAL if data is invalid */
COB_EXPIMP int		cob_put_field_str (const cob_field *, const char *);

// for alphanumeric/national datatypes you may just want to directly use `cob_field->data` up to `cob_field->size` to not need extra buffers
```

This would allow direct "get me the data from COBOL (whatever type and size) into this buffer" and "put this buffer into the data (in the exact format needed)", libcob will handle all the necessary conversions with guaranteed correct values and "as fast as all `MOVE`s are". This would also allow to support literals and other constants (obviously as source only) and special fields like `PIC X ANY LENGTH` or `PIC 9 ANY NUMERIC` (which may be specified in `LINKAGE` and then be of an unknown length and type during compile time).

**Comment by mridoni** [comment]
Yes, this would be nice, even if the parser part would need to be preserved for Gix-IDE (it is used to handle data display and inspection). On the other hand, for this purpose it would also be nice to have a mechanism to dump/parse the AST generated by GnuCOBOL. I know it can be navigated by  using libcob and "following" lists and pointers but a direct (and "compiler-sanctioned") interface would be preferable for using it in tools.

**Comment by GitMensch** [comment]
That would be another GC specific :-) But I agree that this is useful - patches welcomed...

The most AST you currently get on the COBOL fields is what you get when you set/export `COBC_GEN_DUMP_COMMENTS` and then check the generated dump code (this is what cbl-gdb does to get to the symbols), it is a 3.1+ only feature, extended for 3.2 to also contain references to the conditionals (level 88). If you haven't inspected this I highly suggest to do so (and if there's anything missing it could be added there).

**Comment by mridoni** [comment]
Thanks, will surely do.
